modules/apps/oauth2-provider/oauth2-provider-rest/src/main/java/com/liferay/oauth2/provider/rest/internal/endpoint/dynamic/registration/LiferayDynamicRegistrationService.java

```diff
@@ -0,0 +1,282 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2025 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.oauth2.provider.rest.internal.endpoint.dynamic.registration;
+
+import com.liferay.oauth2.provider.rest.internal.endpoint.dynamic.registration.model.LiferayClientRegistration;
+import com.liferay.oauth2.provider.rest.internal.endpoint.dynamic.registration.model.LiferayClientRegistrationResponse;
+import com.liferay.oauth2.provider.util.OAuth2SecureRandomGenerator;
+import com.liferay.portal.kernel.util.HashMapBuilder;
+import com.liferay.portal.kernel.util.ListUtil;
+import com.liferay.portal.kernel.util.Validator;
+import com.liferay.portal.remote.cors.annotation.CORS;
+
+import jakarta.ws.rs.Consumes;
+import jakarta.ws.rs.POST;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.MediaType;
+import jakarta.ws.rs.core.Response;
+import jakarta.ws.rs.core.UriBuilder;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.cxf.jaxrs.utils.ExceptionUtils;
+import org.apache.cxf.jaxrs.utils.JAXRSUtils;
+import org.apache.cxf.rs.security.oauth2.common.Client;
+import org.apache.cxf.rs.security.oauth2.common.OAuthError;
+import org.apache.cxf.rs.security.oauth2.services.ClientRegistration;
+import org.apache.cxf.rs.security.oauth2.services.DynamicRegistrationService;
+import org.apache.cxf.rs.security.oauth2.utils.OAuthUtils;
+
+/**
+ * @author Jorge García Jiménez
+ */
+@Path("/register")
+public class LiferayDynamicRegistrationService
+	extends DynamicRegistrationService {
+
+	@Consumes(MediaType.APPLICATION_JSON)
+	@CORS(allowMethods = "POST")
+	@POST
+	@Produces(MediaType.APPLICATION_JSON)
+	public Response register(LiferayClientRegistration clientRegistration) {
+		return super.register(clientRegistration);
+	}
+
+	@Override
+	protected void checkRegistrationAccessToken(Client c, String accessToken) {
+	}
+
+	protected void fromClientRegistrationToClient(
+		ClientRegistration request, Client client) {
+
+		client.setApplicationName(request.getClientName());
+
+		List<String> grantTypes = client.getAllowedGrantTypes();
+
+		_checkValidGrantResponseTypes(grantTypes, request.getResponseTypes());
+
+		List<String> redirectUris = request.getRedirectUris();
+
+		if (redirectUris != null) {
+			String appType = request.getApplicationType();
+
+			if (appType == null) {
+				appType = "web";
+			}
+
+			for (String redirectUri : redirectUris) {
+				validateRequestUri(redirectUri, appType, grantTypes);
+			}
+
+			client.setRedirectUris(redirectUris);
+		}
+
+		if (client.getRedirectUris(
+			).isEmpty() &&
+			(grantTypes.contains("authorization_code") ||
+			 grantTypes.contains("implicit"))) {
+
+			OAuthError error = new OAuthError(
+				"invalid_request", "A Redirection URI is required");
+
+			_reportInvalidRequestError(error);
```

Comment:
Should we `return` after reporting this error?

Reply:
the _report throws an exception, we can add a return but the flow will be interrupted at that point with or without the return.